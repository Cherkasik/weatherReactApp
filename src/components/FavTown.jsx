import React from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';
import InfoBlock from './InfoBlock';
import Preloader from './Preloader';
import { rmFavourite } from '../favourites/actions';
import { getWeatherByName } from '../weather/actions';
import * as weatherSelector from '../weather/selectors';

const Header = styled.div`
  display: flex;
  justify-content: space-between;
  height: 46px;
`;

const TownContainer = styled.div`
  width: 45%;
  margin-bottom: 20px; 
`;

const Name = styled.div`
  font-size: 26pt;
  font-weight: 700;
  color: #71848B;
`;

const Temperature = styled.div`
  font-size: 26pt;
  color: #71848B;
  display: ${props => props.show ? 'inherit' : 'none'};
`;

const Icon = styled.div`
  width: 50px;
  height: 40px;
  float: left;
  border: 3px solid #71848B;
  text-align: center;
  color: #6E8084;
  font-size: 10pt;
  display: ${props => props.show ? 'inherit' : 'none'};
  
  img {
    width: 50px;
    height: 40px;
  }
`;

const Close = styled.button`
  transform: rotate(45deg);
  border: none;
  background-color: #708189;
  color: #ffffff;
  border-radius: 50%;
  padding: 0 14px;
  font-size: 25pt;
`;

class FavTown extends React.Component {
  constructor(props) {
    super(props);
    this.state = { loading: true };
  }

  componentDidMount() {
    this.props.getWeatherByName(
          this.props.name,
          () => this.setState({ loading: false }));
  }

  render() {
    const { weatherArray, name, rmFavourite } = this.props;
    const { loading } = this.state;
    const weather = weatherArray && weatherArray[name];
    return (
      <TownContainer>
        <Header>
          <Name>{name}</Name>
          <Temperature show={!loading}>{weather && Math.round(weather.main.temp - 273)}Â°C</Temperature>
          <Icon show={!loading}>
            <img src={weather && `https://openweathermap.org/img/w/${weather.weather[0].icon}.png`}  alt="icon"/>
          </Icon>
          <Close onClick={() => rmFavourite(name)}>+</Close>
        </Header>
        {loading
          ? <Preloader small/>
          : <InfoBlock weather={weather} />}
      </TownContainer>
    );
  }
}

const mapStateToProps = state => ({
  weatherArray: weatherSelector.getWeatherNamed(state),
});

const mapDispatchToProps = {
  rmFavourite,
  getWeatherByName,
};

export default connect(mapStateToProps, mapDispatchToProps)(FavTown);
