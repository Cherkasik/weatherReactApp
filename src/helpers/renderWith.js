import React from 'react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import reducer from '../store/rootReducer';

export function renderWithRedux(
  ui,
  { initialState, store = createStore(reducer, initialState) } = {}
) {
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
    store,
  }
}
