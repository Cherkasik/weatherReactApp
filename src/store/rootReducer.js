import { combineReducers } from 'redux';
import geolocation from '../geolocation/reducer';
import favourites from '../favourites/reducer';
import weather from '../weather/reducer';

const appReducer = combineReducers({
  geolocation,
  favourites,
  weather,
});

const rootReducer = (state, actionValue) => {
  return appReducer(state, actionValue);
};

export default rootReducer;