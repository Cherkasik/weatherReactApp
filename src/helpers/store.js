import thunkMiddleware from "redux-thunk";
import {applyMiddleware, compose, createStore} from "redux";
import rootReducer from "../store/rootReducer";

const middlewares = [thunkMiddleware];

if (process.env.NODE_ENV === 'development') {
  const { logger } = require('redux-logger');
  middlewares.push(logger);
}

export const store = compose(applyMiddleware(...middlewares))(createStore)(rootReducer);