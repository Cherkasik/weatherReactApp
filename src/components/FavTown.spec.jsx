import React from 'react';
import 'jest-styled-components';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import FavTown from './FavTown';
import { store } from '../helpers/store';

// TODO: test different combinations of error and loading
describe('FavTown component', () => {
  it('should render correctly', async () => {
    const { container } = render(
      <Provider store={store}>
        <FavTown />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});