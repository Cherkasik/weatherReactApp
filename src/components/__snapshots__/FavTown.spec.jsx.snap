// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavTown component should render correctly 1`] = `
.c6 {
  width: 100%;
  text-align: center;
  margin-top: 20px;
}

.c7 {
  font-size: 25px;
  text-align: center;
  color: #6F8185;
}

.c8 {
  width: 250px;
  height: 80px;
  margin-top: 20px;
  background-color: #C9D2D7;
  border: none;
}

.c8 svg {
  width: 50px;
  height: 50px;
  fill: #6F8185;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  height: 46px;
}

.c0 {
  width: 45%;
  margin-bottom: 20px;
}

.c2 {
  font-size: 26pt;
  font-weight: 700;
  color: #71848B;
}

.c3 {
  font-size: 26pt;
  color: #71848B;
  display: none;
}

.c4 {
  width: 50px;
  height: 40px;
  float: left;
  border: 3px solid #71848B;
  text-align: center;
  color: #6E8084;
  font-size: 10pt;
  display: none;
}

.c4 img {
  width: 50px;
  height: 40px;
}

.c5 {
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  border: none;
  background-color: #708189;
  color: #ffffff;
  border-radius: 50%;
  padding: 0 14px;
  font-size: 25pt;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    />
    <div
      class="c3"
    >
      °C
    </div>
    <div
      class="c4"
    >
      <img
        alt="icon"
      />
    </div>
    <button
      class="c5"
    >
      +
    </button>
  </div>
  <div
    class="c6"
  >
    <div
      class="c7"
    >
      Подождите, данные загружаются
    </div>
    <button
      class="c8"
    >
      <svg>
        Refresh.svg
      </svg>
    </button>
  </div>
</div>
`;
